src/transformers/pipelines/text_to_speech.py

```diff
@@ -0,0 +1,173 @@
+                """Must pass a valid vocoder to the TTSPipeline if speecht5 is used.
+                              Try passing a repo_id or an instance of SpeechT5HifiGan."""
+            )
+
+        self.processor = processor
+        self.vocoder = vocoder
+
+        if self.is_speecht5:
+            self.sampling_rate = vocoder.config.sampling_rate
+        elif sampling_rate is None:
+            # get sampling_rate from config and generation config
+            self.sampling_rate = None
+
+            config = self.model.config.to_dict()
+            gen_config = self.model.__dict__.get("generation_config", None)
+            if gen_config is not None:
+                config.update(gen_config.to_dict())
+
+            for sampling_rate_name in ["sample_rate", "sampling_rate"]:
+                sampling_rate = config.get(sampling_rate_name, None)
+                if sampling_rate is not None:
+                    self.sampling_rate = sampling_rate
+        else:
+            self.sampling_rate = sampling_rate
+
+    def preprocess(self, text, speaker_embeddings=None, **kwargs):
+        if self.is_speecht5:
+            inputs = self.processor(text=text, return_tensors="pt")
+            if speaker_embeddings is None:
+                embeddings_dataset = load_dataset("Matthijs/cmu-arctic-xvectors", split="validation")
+                speaker_embeddings = torch.tensor(embeddings_dataset[7305]["xvector"]).unsqueeze(0)
+
+            return {"input_ids": inputs["input_ids"], "speaker_embeddings": speaker_embeddings}
+
+        processor_args_dict = {}
+
+        SPEAKER_EMBEDDINGS_KEY_MAPPING.get(self.model_type, "speaker_embeddings")
+        if self.model.config.model_type == "bark":
+            # bark speaker embeddings is passed as voice_preset in its processor
+            processor_args_dict["voice_preset"] = speaker_embeddings
+        else:
+            processor_args_dict["speaker_embeddings"] = speaker_embeddings
+
+        output = self.processor(text, **processor_args_dict, **kwargs)
+
+        return output
+
+    def _forward(self, model_inputs, **kwargs):
+        if self.is_speecht5:
+            inputs = model_inputs["input_ids"]
+
+            speaker_embeddings = model_inputs["speaker_embeddings"]
+
+            with torch.no_grad():
+                speech = self.model.generate_speech(inputs, speaker_embeddings, vocoder=self.vocoder)
+        else:
+            if self.only_one_speaker_embeddings:
+                speaker_embeddings_key = SPEAKER_EMBEDDINGS_KEY_MAPPING.get(self.model_type, "speaker_embeddings")
+
+                # check batch_size > 1
+                if len(model_inputs["input_ids"]) > 1 and model_inputs.get(speaker_embeddings_key, None) is not None:
+                    model_inputs[speaker_embeddings_key] = model_inputs[speaker_embeddings_key][0]
+
+            with torch.no_grad():
+                speech = self.model.generate(**model_inputs, **kwargs)
+
+        return speech
```

Comment:
```suggestion
        speech = self.model.generate(**model_inputs)
        return speech
```


`torch.no_grad` is already done for you.
You cannot really inspect the tensors, since batching might be occurring for you.
`self.vocoder` is not allowed (since self isn't).

This function should be model agnostic. Every model that implements `AutoModelForAudioXX` (or something along those lines) should have the same API, and that's what the pipeline should rely on.

Reply:
Hi @Narsil ,
regarding:

> You cannot really inspect the tensors, since batching might be occurring for you.

Batching is taken care for us but it forces additional inputs (for example, `speaker_embeddings`) to be repeated, which is not the expected behavior.
In that specific case, you have to inspect the tensors to know if there is batching.

Reply:
This can boil down to one if/else depending on if the model does auto-regressive generation, or one forward pass:

```python
if self.generate:
    speech = self.model.generate(**model_inputs)
else:
    speech = self.model(**model_inputs)
```

Let's try and handle the vocoder in the SpeechT5 modelling code, and the speaker embeddings outside of this function (e.g. in the `processor`)